File: foodmate-application/src/main/java/com/foodmate/application/food/service/FoodLogService.java
package com.foodmate.application.food.service;

import com.foodmate.shared.food.enums.MealType;

import java.math.BigDecimal;
import java.time.Instant;
import java.util.List;

/** 手工饮食记录第一纵向切片用例。 */
public interface FoodLogService {
    FoodLogView create(long userId, CreateCommand command);

    /** 使用乐观并发完整替换已有饮食记录的可编辑内容。 */
    FoodLogView update(long userId, long foodLogId, long revision, UpdateCommand command);

    List<FoodLogView> list(long userId, Instant from, Instant to);

    /** 查询当前用户可恢复的软删除饮食记录。 */
    List<FoodLogView> listDeleted(long userId);

    void delete(long userId, long foodLogId, long revision, String idempotencyKey);

    FoodLogView restore(long userId, long foodLogId, long revision, String idempotencyKey);

    record CreateCommand(
            Long sessionId,
            Long agentRunId,
            Instant mealTime,
            MealType mealType,
            String notes,
            String idempotencyKey,
            String source,
            Long mealPlanMealId,
            Long compositeDishId,
            Long compositeDishRevision,
            BigDecimal compositeDishServings,
            List<ItemCommand> items) {
        public CreateCommand(
                Long sessionId,
                Long agentRunId,
                Instant mealTime,
                MealType mealType,
                String notes,
                String idempotencyKey,
                List<ItemCommand> items) {
            this(
                    sessionId,
                    agentRunId,
                    mealTime,
                    mealType,
                    notes,
                    idempotencyKey,
                    "manual",
                    null,
                    null,
                    null,
                    null,
                    items);
        }

        public CreateCommand(
                Long sessionId,
                Long agentRunId,
                Instant mealTime,
                MealType mealType,
                String notes,
                String idempotencyKey,
                String source,
                List<ItemCommand> items) {
            this(
                    sessionId,
                    agentRunId,
                    mealTime,
                    mealType,
                    notes,
                    idempotencyKey,
                    source,
                    null,
                    null,
                    null,
                    null,
                    items);
        }

        /** R2 复合菜调用方的兼容构造函数；计划餐次关联使用完整构造函数。 */
        public CreateCommand(
                Long sessionId,
                Long agentRunId,
                Instant mealTime,
                MealType mealType,
                String notes,
                String idempotencyKey,
                String source,
                Long compositeDishId,
                Long compositeDishRevision,
                BigDecimal compositeDishServings,
                List<ItemCommand> items) {
            this(
                    sessionId,
                    agentRunId,
                    mealTime,
                    mealType,
                    notes,
                    idempotencyKey,
                    source,
                    null,
                    compositeDishId,
                    compositeDishRevision,
                    compositeDishServings,
                    items);
        }

        public CreateCommand {
            source = source == null || source.isBlank() ? "manual" : source;
            items = items == null ? List.of() : List.copyOf(items);
        }
    }

    record UpdateCommand(
            Instant mealTime,
            MealType mealType,
            String notes,
            String idempotencyKey,
            Long mealPlanMealId,
            Long compositeDishId,
            Long compositeDishRevision,
            BigDecimal compositeDishServings,
            List<ItemCommand> items) {
        public UpdateCommand(
                Instant mealTime,
                MealType mealType,
                String notes,
                String idempotencyKey,
                List<ItemCommand> items) {
            this(mealTime, mealType, notes, idempotencyKey, null, null, null, null, items);
        }

        /** R2 复合菜调用方的兼容构造函数。 */
        public UpdateCommand(
                Instant mealTime,
                MealType mealType,
                String notes,
                String idempotencyKey,
                Long compositeDishId,
                Long compositeDishRevision,
                BigDecimal compositeDishServings,
                List<ItemCommand> items) {
            this(
                    mealTime,
                    mealType,
                    notes,
                    idempotencyKey,
                    null,
                    compositeDishId,
                    compositeDishRevision,
                    compositeDishServings,
                    items);
        }

        public UpdateCommand {
            items = items == null ? List.of() : List.copyOf(items);
        }
    }

    record ItemCommand(String rawName, BigDecimal amount, String unit, Long nutritionFoodId) {
        public ItemCommand(String rawName, BigDecimal amount, String unit) {
            this(rawName, amount, unit, null);
        }
    }

    record FoodLogView(
            long foodLogId,
            Long sessionId,
            Long agentRunId,
            Instant mealTime,
            MealType mealType,
            String notes,
            String source,
            Long mealPlanMealId,
            Long compositeDishId,
            Long compositeDishRevision,
            BigDecimal compositeDishServings,
            long revision,
            boolean deleted,
            Instant createdAt,
            Instant updatedAt,
            List<ItemView> items) {
        public FoodLogView(
                long foodLogId,
                Long sessionId,
                Long agentRunId,
                Instant mealTime,
                MealType mealType,
                String notes,
                String source,
                long revision,
                boolean deleted,
                Instant createdAt,
                Instant updatedAt,
                List<ItemView> items) {
            this(
                    foodLogId,
                    sessionId,
                    agentRunId,
                    mealTime,
                    mealType,
                    notes,
                    source,
                    null,
                    null,
                    null,
                    null,
                    revision,
                    deleted,
                    createdAt,
                    updatedAt,
                    items);
        }

        /** R2 复合菜响应的兼容构造函数。 */
        public FoodLogView(
                long foodLogId,
                Long sessionId,
                Long agentRunId,
                Instant mealTime,
                MealType mealType,
                String notes,
                String source,
                Long compositeDishId,
                Long compositeDishRevision,
                BigDecimal compositeDishServings,
                long revision,
                boolean deleted,
                Instant createdAt,
                Instant updatedAt,
                List<ItemView> items) {
            this(
                    foodLogId,
                    sessionId,
                    agentRunId,
                    mealTime,
                    mealType,
                    notes,
                    source,
                    null,
                    compositeDishId,
                    compositeDishRevision,
                    compositeDishServings,
                    revision,
                    deleted,
                    createdAt,
                    updatedAt,
                    items);
        }

        public FoodLogView {
            items = items == null ? List.of() : List.copyOf(items);
        }
    }

    record ItemView(
            long foodLogItemId,
            int itemOrder,
            String rawName,
            Long nutritionFoodId,
            BigDecimal amount,
            String unit,
            String nutritionStatus,
            BigDecimal caloriesKcal,
            BigDecimal proteinG,
            BigDecimal fatG,
            BigDecimal carbsG) {
        public ItemView(
                long foodLogItemId,
                int itemOrder,
                String rawName,
                BigDecimal amount,
                String unit,
                String nutritionStatus,
                BigDecimal caloriesKcal,
                BigDecimal proteinG,
                BigDecimal fatG,
                BigDecimal carbsG) {
            this(
                    foodLogItemId,
                    itemOrder,
                    rawName,
                    null,
                    amount,
                    unit,
                    nutritionStatus,
                    caloriesKcal,
                    proteinG,
                    fatG,
                    carbsG);
        }
    }
}

File: foodmate-ui/src/services/foodLogService.ts
import { apiRequest } from './apiClient';

export type FoodLogItem = {
  food_log_item_id: string;
  item_order: number;
  raw_name: string;
  nutrition_food_id?: string | null;
  amount: number | string;
  unit: string;
  nutrition_status: string;
  calories_kcal: number | string | null;
  protein_g: number | string | null;
  fat_g: number | string | null;
  carbs_g: number | string | null;
};

export type FoodLog = {
  food_log_id: string;
  meal_time: string;
  meal_type: 'breakfast' | 'lunch' | 'dinner' | 'snack' | string;
  notes: string | null;
  source: string;
  meal_plan_meal_id?: string | null;
  composite_dish_id?: string | null;
  composite_dish_revision?: number | null;
  composite_dish_servings?: number | string | null;
  revision: number;
  deleted: boolean;
  items: FoodLogItem[];
};

export type FoodLogWriteRequest = {
  meal_time: string;
  meal_type: FoodLog['meal_type'];
  notes?: string;
  meal_plan_meal_id?: string;
  composite_dish_id?: string;
  composite_dish_revision?: number;
  composite_dish_servings?: number;
  items: Array<{ raw_name: string; amount: number; unit: string; nutrition_food_id?: string }>;
};

export async function loadFoodLogs(from: string, to: string, signal?: AbortSignal): Promise<FoodLog[]> {
  return apiRequest<FoodLog[]>(
    `/api/food-logs?from=${encodeURIComponent(from)}&to=${encodeURIComponent(to)}`,
    signal ? { signal } : undefined,
  );
}

export async function createFoodLog(request: FoodLogWriteRequest): Promise<FoodLog> {
  return apiRequest<FoodLog>('/api/food-logs', {
    method: 'POST',
    headers: { 'Idempotency-Key': idempotencyKey('food-log-create') },
    body: JSON.stringify(request),
  });
}

export async function updateFoodLog(
  foodLogId: string,
  revision: number,
  request: FoodLogWriteRequest,
): Promise<FoodLog> {
  return apiRequest<FoodLog>(`/api/food-logs/${encodeURIComponent(foodLogId)}?revision=${revision}`, {
    method: 'PATCH',
    headers: { 'Idempotency-Key': idempotencyKey('food-log-update') },
    body: JSON.stringify(request),
  });
}

export async function deleteFoodLog(foodLogId: string, revision: number): Promise<void> {
  await apiRequest<void>(`/api/food-logs/${encodeURIComponent(foodLogId)}?revision=${revision}`, {
    method: 'DELETE',
    headers: { 'Idempotency-Key': idempotencyKey('food-log-delete') },
  });
}

export async function loadDeletedFoodLogs(signal?: AbortSignal): Promise<FoodLog[]> {
  return apiRequest<FoodLog[]>('/api/food-logs/deleted', signal ? { signal } : undefined);
}

export async function restoreFoodLog(foodLogId: string, revision: number): Promise<FoodLog> {
  return apiRequest<FoodLog>(`/api/food-logs/${encodeURIComponent(foodLogId)}/restore?revision=${revision}`, {
    method: 'POST',
    headers: { 'Idempotency-Key': idempotencyKey('food-log-restore') },
  });
}

function idempotencyKey(prefix: string) {
  const suffix = globalThis.crypto?.randomUUID?.() ?? `${Date.now()}-${Math.random().toString(16).slice(2)}`;
  return `${prefix}-${suffix}`;
}
